src/houses/KM_HouseBarracks.pas

```diff
@@ -277,11 +279,24 @@ procedure TKMHouseBarracks.CreateRecruitInside(aIsMapEd: Boolean);
 end;
 
 
+procedure TKMHouseBarracks.ValidateNUpdateRallyPoint;
+begin
+  //this will automatically update rally point to valid value
+  SetRallyPoint(fRallyPoint);
+end;
+
+
+procedure TKMHouseBarracks.SetRallyPoint(aRallyPoint: TKMPoint);
+begin
+  fRallyPoint := gTerrain.GetPassablePointWithinSegment(PointBelowEntrance, aRallyPoint, tpWalk);
+end;
+
+
 function TKMHouseBarracks.GetRallyPoint: TKMPoint;
 begin
   if not gTerrain.CheckPassability(fRallyPoint, tpWalk) then
     //update rally point if its not valid (not walkable). Set it closer to barracks entrance (PointBelowEntrance)
-    fRallyPoint := gTerrain.GetPassablePointWithinSegment(PointBelowEntrance, fRallyPoint, tpWalk);
+    SetRallyPoint(fRallyPoint);
```

Comment:
Unrelated to this pull request (I only just noticed), but what you are doing here is unsafe for replays/multiplayer. GetRallyPoint should be side-effect free, since it will be called non-deterministically by the GUI when the player selects the house. This will cause multiplayer clients and replays to go out of sync when the previous rally point becomes unwalkable and the player selects the house.

Can you please submit a pull request to fix this? I suggest GetRallyPoint just returns fRallyPoint (so could be unwalkable), but then when you use  the rally point (when soldier is trained or woodcutter goes to work) you call ValidateRallyPoint first. This will be from within the game code so safe for replays/MP.